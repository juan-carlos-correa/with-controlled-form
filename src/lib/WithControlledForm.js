import React, { Component } from 'react';
import { Validators } from './Validators';

export const WithControlledForm = (FormComponent, state = {}, formValidations = {}) => (
  class WithFormMethodsHOC extends Component {
    constructor (props) {
      super(props);

      this.state = {
        values: { ...state },
        errors: this._stateToErrors(state),
      };
    }

    _stateToErrors = (state) => {
      const errors = {};

      const keys = Object.keys(state);
      for (let i = 0; i < keys.length; i += 1) {
        errors[keys[i]] = [];
      }

      return errors;
    }

    _validateForm = (values) => {
      const result = Validators.validate(values, formValidations);
      this.setState({ errors: result.errors });
      return result.isValid;
    }

    isFormClean = () => {
      const { values } = this.state;
      const isValue = Object.values(values).filter(val => val.length > 0);
      return isValue.length === 0;
    }

    cleanForm = () => {
      const newValues = {};
      const { values } = this.state;

      for (let val in values) {
        newValues[val] = '';
      }

      this.setState({
        values: newValues,
        errors: this._stateToErrors(newValues),
      });
    }

    handleChange = (e) => {
      const { name, value } = e.target;
      const { values } = this.state;

      values[name] = value;
      this.setState({ values });
    }

    handleCheckboxChange = (e) => {
      const { name, checked } = e.target;
      const { values } = this.state;

      values[name] = checked;
      this.setState({ values });
    }

    handleBlur = (e) => {
      e.preventDefault();

      const { errors } = this.state;
      const { name, value } = e.target;
      const validations = formValidations[name];
      const result = Validators.validateOne(value, validations);
      errors[name] = !result.isValid ? result.errors[0] : [];
      this.setState({ errors });
    }

    handleSubmit = (e) => {
      e.preventDefault();
      const { values } = this.state;
      const { handleSubmit } = this.props;

      if (this._validateForm(values)) {
        handleSubmit(values);
      }
    }

    render () {
      return (
        <FormComponent
          {...this.props}
          {...this.state}
          cleanForm={this.cleanForm}
          isFormClean={this.isFormClean}
          handleChange={this.handleChange}
          handleCheckboxChange={this.handleCheckboxChange}
          handleBlur={this.handleBlur}
          handleSubmit={this.handleSubmit}
        />
      )
    }
  }
);
