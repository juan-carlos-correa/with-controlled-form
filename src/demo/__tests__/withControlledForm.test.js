import React from 'react';
import { mount } from 'enzyme';
import Form from '../components/Form';
import WithControlledForm from '../../lib/WithControlledForm';

const formState = {
  email: '',
  phoneNumber: '',
};

const formValidations = {
  email: {
    isRequired: true,
    isMinLength: 3,
    isMaxLength: 60,
    isEmail: true,
  },
  phoneNumber: {
    isRequired: true,
    isMinLength: 6,
    isMaxLength: 15,
  }
};

test('WithControlledForm render', () => {
  const FormWithControlled = WithControlledForm(Form, formState, formValidations);
  const wrapper = mount(<FormWithControlled />);
  expect(wrapper).toMatchSnapshot();

  wrapper.unmount();
});

test('WithControlledForm state tests', () => {
  const handleSubmitMock = jest.fn();
  const FormWithControlled = WithControlledForm(Form, formState, formValidations);
  const wrapper = mount(<FormWithControlled handleSubmit={handleSubmitMock} />);

  let stateToTest = { values: formState, errors: formState };
  let componentState = null;

  // initial state
  componentState = wrapper.instance().state;
  expect(componentState).toEqual(stateToTest);

  // update state with events
  wrapper.find('#email').simulate('change', { target: { value: 'foo', name: 'email' } });
  stateToTest = { ...stateToTest, values: { ...stateToTest.values, email: 'foo' } };
  componentState = wrapper.instance().state;
  expect(componentState).toEqual(stateToTest);

  wrapper.find('#phoneNumber').simulate('change', { target: { value: '123456', name: 'phoneNumber' } });
  stateToTest = { ...stateToTest, values: { ...stateToTest.values, phoneNumber: '123456' } };
  componentState = wrapper.instance().state;
  expect(componentState).toEqual(stateToTest);

  wrapper.find('#send').simulate('submit', {})
  stateToTest = { ...stateToTest, errors: { ...stateToTest.errors, email: 'Este valor debe ser un email del tipo name@mail.com' } };
  componentState = wrapper.instance().state;
  expect(componentState).toEqual(stateToTest)

  // update state errors
  wrapper.find('#email').simulate('change', { target: { value: 'name@mail.com', name: 'email' } });
  stateToTest = {
    ...stateToTest,
    values: {
      ...stateToTest.values,
      email: 'name@mail.com'
    },
    errors: {
      phoneNumber: '',
      email: ''
    }
  };
  wrapper.find('#send').simulate('submit', {})
  componentState = wrapper.instance().state;
  expect(componentState).toEqual(stateToTest);

  // handleSubmitMock should be called on submit with valid form data
  expect(handleSubmitMock.mock.calls.length).toBe(1);

  wrapper.unmount();
})
