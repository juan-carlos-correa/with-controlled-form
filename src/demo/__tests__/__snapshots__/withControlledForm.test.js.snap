// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithControlledForm render 1`] = `
<WithFormMethodsHOC>
  <Form
    cleanForm={[Function]}
    errors={
      Object {
        "email": Array [],
        "phoneNumber": Array [],
      }
    }
    handleBlur={[Function]}
    handleChange={[Function]}
    handleFocus={[Function]}
    handleSubmit={[Function]}
    isFormClean={[Function]}
    values={
      Object {
        "email": "",
        "phoneNumber": "",
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-element"
      >
        <label
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="input"
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="form-element"
      >
        <label
          htmlFor="phoneNumber"
        >
          Phone number:
        </label>
        <input
          className="input"
          id="phoneNumber"
          name="phoneNumber"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="tel"
          value=""
        />
        <div
          className="error-message"
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="button"
          id="send"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </Form>
</WithFormMethodsHOC>
`;
